### PR TITLE
[BEAM-2502, BEAM-2504, BEAM-2505] Fix processing time timer handling in ReduceFnRunner

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

A few major correctness issues.

R: @bjchambers

### CODE REVIEW

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java [hunk 1]
```diff
@@ -351,6 +389,38 @@ public BoundedWindow window() {
   }
 
   /**
+   * Tests that when a processing time timers comes in after a window is expired
```
Comment (bjchambers) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Is it worth having one that tests that when a processing time timer comes in with a timestamp > GC time, it doesn't *cause* GC (mostly asking given that was one of the bugs, that isn't explicitly tested here).

Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Oh, actually I think my comment is not clear. That is what is tested.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java [hunk 2]
```diff
@@ -140,7 +141,40 @@ public Void answer(InvocationOnMock invocation) throws Exception {
...
     })
     .when(mockTrigger).onFire(anyTriggerContext());
- }
+  }
+
+  /**
+   * Tests that a processing time timers does not cause window GC.
+   */
+  @Test
+  public void testProcessingTimeTimersDoesNotGc() throws Exception {
```
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: @bjchambers some grammar errors... but this one fires  a processing time timer with a timestamp long after the end of the window, but no watermark advancement. This test does fail without the change to `ReduceFnRunner`, confirming that it definitely causes a GC.

Comment (bjchambers) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/ReduceFnRunnerTest.java]: Ah, LGTM

**Pull request merged**